File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.java
class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        int n = nums.length, ans=-1;
        long low,high,mid,target;
        low = 1;
        high = max(nums,n);
        while(low<=high){
            mid = (low+high)/2;
            target = check(nums,mid,n);
            if(target <= threshold){
                ans = (int)mid;
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        return ans;
    }
    public int check(int[] nums,long mid,int n){
        int result = 0;
        double val;
        for(int i=0;i<n;i++){
            val = (double)nums[i]/(double)mid;
            result += (int) Math.ceil(val);
        }
        return result;
    }
    public int max(int[] nums,int n){
        int maxi = Integer.MIN_VALUE;
        for(int i=0;i<n;i++){
            if(maxi < nums[i])
                maxi = nums[i];
        }
        return maxi;
    }
}

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.go
func smallestDivisor(nums []int, threshold int) int {
    n := len(nums)
    ans := -1
    low := 1
    high := max(nums,n)
    var mid,target int
    for low<=high{
        mid = (low+high)/2
        target = check(nums,mid,n)
        if target <= threshold{
            ans = mid
            high = mid-1
        }else{
            low = mid+1
        }
    }
    return ans;
}

func check(nums []int,mid int,n int) int{
    result := 0
    var val float64
    for i:=0;i<n;i++{
        val = float64(nums[i])/float64(mid)
        result = result + int(math.Ceil(val))
    }
    return result
}

func max(nums []int,n int) int{
    maxi := math.MinInt64
    for i:=0;i<n;i++{
        if(maxi < nums[i]){
            maxi = nums[i]
        }
    }
    return maxi
}